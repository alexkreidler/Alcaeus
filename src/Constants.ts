export const JsonLd = {
    Graph: '@graph',
    Context: '@context',
    Id: '@id',
    Value: '@value',
    Type: '@type',
    Language: '@language',
}

export const MediaTypes = {
    jsonLd: 'application/ld+json',
    ntriples: 'application/n-triples',
    nquads: 'application/n-quads',
}

export const Headers = {
    Link: 'Link',
    Location: 'Location',
    ContentType: 'Content-Type',
}
