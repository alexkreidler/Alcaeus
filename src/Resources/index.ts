import type { RdfResource } from '@tpluscode/rdfine'
import type { DatasetCore } from 'rdf-js'
import type { HydraResponse } from '../alcaeus'
import type { OperationFinder } from './CoreMixins/OperationFinder'
import type { ManagesBlockPattern } from './Mixins/ManagesBlock'
import type { Operation } from './Operation'
import type { SupportedProperty } from './Mixins/SupportedProperty'

export interface HydraResource<D extends DatasetCore = DatasetCore> extends RdfResource<D>, OperationFinder {
    /**
     * Gets the operations which can be performed on this resource
     */
    readonly operations: Operation[]

    /**
     * Gathers all properties from current resource's classes
     */
    getProperties(): { supportedProperty: SupportedProperty; objects: any[] }[]

    /**
     * Get all property/value pairs for hydra:Link properties
     *
     * @param includeMissing if true, will include properties not present in resource representation
     */
    getLinks(includeMissing?: boolean): { supportedProperty: SupportedProperty; resources: HydraResource[]}[]

    /**
     * Gets objects of hydra:collection property
     */
    getCollections(filter?: ManagesBlockPattern): HydraResource[]

    /**
     * Dereferences the resource
     */
    load?(): Promise<HydraResponse<this>>
}

export interface View extends HydraResource {
    /**
     * Gets the actual collection resource, of which this view is part
     */
    readonly collection: HydraResource | null
}

export type { DocumentedResource } from './Mixins/DocumentedResource'
export type { Class } from './Mixins/Class'
export type { SupportedOperation } from './Mixins/SupportedOperation'
export type { SupportedProperty } from './Mixins/SupportedProperty'
export type { Collection } from './Mixins/Collection'
export type { PartialCollectionView } from './Mixins/PartialCollectionView'
export type { ApiDocumentation } from './Mixins/ApiDocumentation'
