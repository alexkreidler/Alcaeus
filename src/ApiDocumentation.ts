/// <reference path="../typings/browser.d.ts" />
'use strict';

//noinspection TypeScriptCheckImport
import {promises as jsonld} from 'jsonld';
import * as _ from 'lodash';
import {Core, JsonLd} from './Constants';

export class ApiDocumentation {
    private _original;

    constructor(apiDoc:any) {
        this._original = apiDoc;
    }

    static load(uri:string):Promise<ApiDocumentation> {
        return null;
    }

    getOperations(classUri:string):Promise<Array<Operation>> {
        return jsonld.flatten(this._original, Core.Context)
            .then(flat => {
                var supportedClass = _.find(flat[JsonLd.Graph], obj => obj._id === classUri);

                return _.chain(flat[JsonLd.Graph])
                    .filter(obj => obj._id === supportedClass.supportedOperation || _.some(supportedClass.supportedOperation, sp => sp === obj._id))
                    .map(op => new Operation(op))
                    .value();
            });
    }
}

export class Operation {
    constructor(hydraOperation:any) {
    }

    get description() {
        return 'Gets the api#Class';
    }

    get method() {
        return 'GET';
    }
}
