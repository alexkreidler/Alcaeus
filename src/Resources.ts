'use strict';
import * as _ from 'lodash';
import {promises as jsonld} from 'jsonld';
import { nonenumerable } from 'core-decorators';
import {JsonLd, Core} from './Constants';
//noinspection TypeScriptCheckImport
import {default} from 'core-js/es6/weak-map';

var _isProcessed = new WeakMap();

export class Resource implements IResource {

    constructor(actualResource) {
        _.extend(this, actualResource);

        _isProcessed.set(this, false);
    }

    @nonenumerable
    get id() {
        return this[JsonLd.Id];
    }

    @nonenumerable
    get types() {
        var types = this[JsonLd.Type];

        if(typeof types === 'string'){
            return [ types ];
        }

        return types;
    }

    @nonenumerable
    get _processed() {
        return _isProcessed.get(this);
    }

    @nonenumerable
    set _processed(val:boolean) {
        _isProcessed.set(this, val);
    }

    compact(context:any = null) {
        return jsonld.compact(this, context || Core.Context);
    }
}

var _apiDocumentation = new WeakMap();
var _incomingLinks = new WeakMap();

export class HydraResource extends Resource implements IHydraResource {

    constructor(heracles:IHeracles, actualResource, apiDoc:IApiDocumentation, incomingLinks) {
        super(actualResource);

        _apiDocumentation.set(this, apiDoc);
        _incomingLinks.set(this, incomingLinks);
    }

    @nonenumerable
    get apiDocumentation() {
        return _apiDocumentation.get(this);
    }

    getIncomingLinks() {
        return _incomingLinks.get(this);
    }

    @nonenumerable
    get operations() {
        var classOperations;
        if(_.isArray(this[JsonLd.Type])) {
            classOperations = _.map(this[JsonLd.Type], type => this.apiDocumentation.getOperations(type));
        } else {
            classOperations = [ this.apiDocumentation.getOperations(this[JsonLd.Type]) ];
        }

        var propertyOperations = _.chain(this.getIncomingLinks())
            .map(link => this.apiDocumentation.getOperations(link[0], link[1]))
            .union()
            .value();

        var operations = [...classOperations, ...propertyOperations];
        return _.flatten(operations);
    }
}

export class PartialCollectionView extends HydraResource implements IPartialCollectionView {

    @nonenumerable
    get first() { return this[Core.Vocab.first] || null; }

    @nonenumerable
    get previous() { return this[Core.Vocab.previous] || null; }

    @nonenumerable
    get next() { return this[Core.Vocab.next] || null; }

    @nonenumerable
    get last() { return this[Core.Vocab.last] || null; }

    @nonenumerable
    get collection() {
        var collectionLink = _.find(this.getIncomingLinks(), linkArray => {
            return linkArray.predicate === Core.Vocab.view
        });

        return collectionLink ? collectionLink.subject : null
    }
}