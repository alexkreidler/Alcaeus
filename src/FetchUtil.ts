'use strict';

//noinspection TypeScriptCheckImport
import * as li from 'li';
//noinspection TypeScriptCheckImport
import {promises as jsonld} from 'jsonld';
import {ApiDocumentation} from "./ApiDocumentation";
import * as Constants from "./Constants";

export class FetchUtil {
    static fetchResource(uri:string, fetchApiDocs:boolean = true):Promise<ExpandedWithDocs> {
        var requestAcceptHeaders = Constants.MediaTypes.jsonLd + ', ' + Constants.MediaTypes.ntriples + ', ' + Constants.MediaTypes.nquads;

        return window.fetch(uri, <FetchOptions>{
                headers: {
                    accept: requestAcceptHeaders
                }
            })
            .then((res:Response) => {
                var apiDocsPromise;
                if (fetchApiDocs) {
                    apiDocsPromise = fetchDocumentation(res);
                } else {
                    apiDocsPromise = Promise.resolve(null);
                }

                return Promise.all([getJsObject(res), apiDocsPromise])
                    .then(values => {
                        return new ExpandedWithDocs(values[0], values[1]);
                    });
            })
    }
}

function fetchDocumentation(res:Response):Promise<ApiDocumentation> {
    if (res.headers.has(Constants.Headers.Link)) {
        var linkHeaders = res.headers.get(Constants.Headers.Link);
        var links = li.parse(linkHeaders);

        if (links[Constants.Core.Vocab.apiDocumentation]) {
            return ApiDocumentation.load(links[Constants.Core.Vocab.apiDocumentation]);
        }
    }

    return Promise.resolve(null);
}

class ExpandedWithDocs {
    constructor(resources:Object, apiDocumentation:ApiDocumentation) {
        this.resources = resources;
        this.apiDocumentation = apiDocumentation;
    }

    resources:Object;
    apiDocumentation:ApiDocumentation
}

class FetchError extends Error {
    private _response;

    constructor(response:Response) {
        super('Request failed');

        this._response = response;
    }

    get response() {
        return this._response;
    }
}

function getJsObject(res:Response) {
    var mediaType = res.headers.get(Constants.Headers.ContentType) || Constants.MediaTypes.jsonLd;

    if(res.ok === false){
        return Promise.reject(new FetchError(res))
    }

    if (mediaType === Constants.MediaTypes.jsonLd) {
        return res.json().then(getFlattenedGraph);
    } else {

        if (mediaType === Constants.MediaTypes.ntriples ||
            mediaType === Constants.MediaTypes.ntriples) {
            mediaType = 'application/nquads';
        }

        return res.text().then(rdf => {
            return jsonld.fromRDF(rdf, {format: mediaType}).then(getFlattenedGraph);
        });
    }
}

function getFlattenedGraph(json) {
    return jsonld.flatten(json, {})
        .then(flattened => flattened[Constants.JsonLd.Graph]);
}