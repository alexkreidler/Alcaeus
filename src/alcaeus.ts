import { ResourceFactory } from '@tpluscode/rdfine'
import $rdf from 'rdf-ext'
import DatasetExt from 'rdf-ext/lib/Dataset'
import * as FetchUtil from './FetchUtil'
import { merge } from './helpers/MergeHeaders'
import { IHydraResponse, create } from './HydraResponse'
import { IMediaTypeProcessor } from './MediaTypeProcessors/RdfProcessor'
import { IOperation } from './Resources'
import { IResponseWrapper } from './ResponseWrapper'
import { IRootSelector } from './RootSelectors'

export interface IHydraClient {
    rootSelectors: IRootSelector[];
    mediaTypeProcessors: { [name: string]: IMediaTypeProcessor };
    loadResource(uri: string, headers?: HeadersInit): Promise<IHydraResponse>;
    invokeOperation(operation: IOperation, uri: string, body?: BodyInit, headers?: string | HeadersInit): Promise<IHydraResponse>;
    defaultHeaders: HeadersInit | (() => HeadersInit);
    dataset: DatasetExt;
    factory: ResourceFactory;
}

const addOrReplaceGraph = async (
    alcaeus: IHydraClient,
    response: IResponseWrapper,
    uri: string): Promise<void> => {
    const suitableProcessor = Object.values(alcaeus.mediaTypeProcessors)
        .find((processor) => processor.canProcess(response.mediaType))

    if (!suitableProcessor) {
        console.warn(`No processor found for media type ${response.mediaType}`)
        return
    }

    const graph = $rdf.namedNode(uri)
    const dataset = await suitableProcessor.process(uri, response)
    await alcaeus.dataset
        .removeMatches(undefined, undefined, undefined, graph)
        .import(dataset.map(quad => {
            return $rdf.quad(
                quad.subject,
                quad.predicate,
                quad.object,
                graph,
            )
        }).toStream())
}

async function getApiDocumentation (this: Alcaeus, response: IResponseWrapper, headers): Promise<void> {
    if (!response.apiDocumentationLink) {
        console.warn(`Resource ${response.requestedUri} does not expose API Documentation link`)
        return
    }

    await this.loadDocumentation(response.apiDocumentationLink, headers)
}

export class Alcaeus implements IHydraClient {
    public rootSelectors: IRootSelector[];

    public mediaTypeProcessors: { [name: string]: IMediaTypeProcessor };

    public defaultHeaders: HeadersInit | (() => HeadersInit) = {}

    public readonly dataset: DatasetExt = $rdf.dataset()

    public readonly factory: ResourceFactory

    public constructor (
        rootSelectors: IRootSelector[],
        mediaTypeProcessors: { [name: string]: IMediaTypeProcessor },
        factory: ResourceFactory
    ) {
        this.rootSelectors = rootSelectors
        this.mediaTypeProcessors = mediaTypeProcessors
        this.factory = factory
    }

    public async loadResource (uri: string, headers: HeadersInit = {}): Promise<IHydraResponse> {
        const response = await FetchUtil.fetchResource(uri, this.__mergeHeaders(new Headers(headers)))
        await addOrReplaceGraph(this, response, uri)
        getApiDocumentation.call(this, response, headers)

        return create(uri, response, this)
    }

    public async loadDocumentation (uri: string, headers: HeadersInit = {}) {
        const response = await FetchUtil.fetchResource(uri, this.__mergeHeaders(new Headers(headers)))
        await addOrReplaceGraph(this, response, uri)
    }

    public async invokeOperation (operation: IOperation, uri: string, body?: BodyInit, headers: string | HeadersInit = {}): Promise<IHydraResponse> {
        if (typeof headers === 'string') {
            headers = {
                'content-type': headers,
            }

            // TODO: remove in 1.0
            console.warn('DEPRECATION NOTICE: passing content type as string will be removed in version 1.0')
        }

        const mergedHeaders = this.__mergeHeaders(new Headers(headers))

        const response = await FetchUtil.invokeOperation(operation.method, uri, body, mergedHeaders)
        getApiDocumentation.call(this, response, headers)

        return create(uri, response, this)
    }

    private __mergeHeaders (headers: Headers): Headers {
        const defaultHeaders = typeof this.defaultHeaders === 'function' ? this.defaultHeaders() : this.defaultHeaders

        return merge(new Headers(defaultHeaders), headers)
    }
}
