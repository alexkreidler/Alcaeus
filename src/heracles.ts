'use strict';
import * as _ from 'lodash';
import {FetchUtil} from './FetchUtil';
import {JsonLd, Core} from './Constants';
import {JsonLdUtil} from "./JsonLdUtil";
import 'rxjs/add/operator/toPromise';
import {ResourceFactory as ResourceFactoryCtor} from './ResourceFactory';
import {HydraResource as ResourceCtor} from "./Resources";
import {IHeracles, IHydraResource, IApiDocumentation, IOperation} from './interfaces';

class Heracles implements IHeracles {
    public resourceFactory = new ResourceFactoryCtor();

    loadResource(uri:string):Promise<IHydraResource> {
        return FetchUtil.fetchResource(uri)
            .then(processFetchUtilResponse.call(this, uri));
    }

    loadDocumentation(uri:string):Promise<IApiDocumentation> {
        return FetchUtil.fetchResource(uri)
            .then(response => {
                var typeOverrides = {};
                typeOverrides[JsonLdUtil.trimTrailingSlash(uri)] = Core.Vocab.ApiDocumentation;

                return getRequestedObject(this, uri, response.resources, typeOverrides)(null);
            }, () => null);
    }

    invokeOperation(operation:IOperation, uri:string, body:any, mediaType?:string):Promise<IHydraResource> {
        return FetchUtil.invokeOperation(operation.method, uri, body, mediaType)
            .then(processFetchUtilResponse.call(this, uri));
    }
}

export var ResourceFactory = ResourceFactoryCtor;
export var Resource = ResourceCtor;
export var Hydra = new Heracles();

function processFetchUtilResponse(uri) {
    return response =>
        this.loadDocumentation(response.apiDocumentationLink)
            .then(getRequestedObject(this, uri, response.resources));
}

function getRequestedObject(heracles:IHeracles, uri, resources, typeOverrides = {}) {
    return apiDocumentation => {
        var resourcified = _.keyBy(resources, res => JsonLdUtil.trimTrailingSlash(res[JsonLd.Id]));
        uri = JsonLdUtil.trimTrailingSlash(uri);

        _.transform(resources, (acc, val) => {
            var id = JsonLdUtil.trimTrailingSlash(val[JsonLd.Id]);
            acc[id] = heracles.resourceFactory.createResource(heracles, val, apiDocumentation, acc, typeOverrides[id]);
        }, resourcified);

        if (!resourcified[uri]) {
            return Promise.reject(new Error('Resource ' + uri + ' was not found in the response'));
        }

        _.each(resourcified, g => resourcify(heracles, g, resourcified, apiDocumentation, typeOverrides));

        return resourcified[uri];
    };
}

function resourcify(heracles, obj, resourcified, apiDoc, typeOverrides) {
    if (_.isObject(obj) === false) {
        return obj;
    }

    if (obj[JsonLd.Value]) {
        return obj[JsonLd.Value];
    }

    var selfId = JsonLdUtil.trimTrailingSlash(obj[JsonLd.Id]);

    if (!selfId) {
        return obj;
    }

    var resource = resourcified[selfId];
    if (!resource || typeof resource._processed === 'undefined') {
        var id = JsonLdUtil.trimTrailingSlash(obj[JsonLd.Id]);
        resource = heracles.resourceFactory.createResource(heracles, obj, apiDoc, resourcified, id);
        resourcified[selfId] = resource;
    }

    if (resource._processed === true) {
        return resource;
    }

    resource._processed = true;
    _.forOwn(resource, (value, key) => {
        if (_.isArray(value)) {
            resource[key] = _.map(value, el => resourcify(heracles, el, resourcified, apiDoc, typeOverrides));
            return;
        }

        resource[key] = resourcify(heracles, value, resourcified, apiDoc, typeOverrides);
    });

    return resource;
}