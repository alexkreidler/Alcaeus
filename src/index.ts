import { EventEmitter } from 'events'
import { ResourceFactoryImpl } from '@tpluscode/rdfine'
import { Sink, Stream } from 'rdf-js'
import { Alcaeus, HydraClient } from './alcaeus'
import * as coreMixins from './Resources/CoreMixins'
import * as mixins from './ResourceFactoryDefaults'
import { AllDefault, RootSelector } from './RootSelectors'
import Resource from './Resources/Resource'

export { Alcaeus } from './alcaeus'
export { default as Resource } from './Resources/Resource'
export { ResourceIdentifier, ResourceIndexer } from '@tpluscode/rdfine'
export * from './Resources/index'
export { Operation } from './Resources/Operation'

export const defaultRootSelectors = Object.values(AllDefault)

interface AlcaeusInit {
    rootSelectors?: RootSelector[];
    parsers?: Record<string, Sink<EventEmitter, Stream>>;
}

export function create ({ rootSelectors, parsers }: AlcaeusInit = {}): HydraClient {
    let factory: ResourceFactoryImpl<HydraResource>
    class HydraResource extends Resource {
        public static get factory () {
            return factory
        }
    }

    factory = new ResourceFactoryImpl(HydraResource)
    const alcaeus = new Alcaeus(rootSelectors || defaultRootSelectors, factory)

    if (parsers) {
        Object.entries(parsers).forEach(pair => alcaeus.parsers.set(...pair))
    }

    factory.addMixin(coreMixins.createResourceLoaderMixin(alcaeus))
    factory.addMixin(coreMixins.createHydraResourceMixin(alcaeus))
    factory.addMixin(coreMixins.OperationFinderMixin)
    Object.values(mixins).forEach(mixin => factory.addMixin(mixin))

    return alcaeus
}

export default create()
