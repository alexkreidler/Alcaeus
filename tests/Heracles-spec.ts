'use strict';
import * as sinon from 'sinon';
//noinspection TypeScriptCheckImport
import {promises as jsonld} from 'jsonld';
import {Hydra} from '../src/heracles';
import {Resource} from '../src/Resources';
import {FetchUtil} from '../src/FetchUtil';
import {JsonLd, Core} from '../src/Constants';
import {Bodies, Documentations} from './test-objects';

describe('Hydra', () => {

    beforeEach(() => {
        sinon.spy(Hydra.resourceFactory, 'createResource');
        sinon.stub(FetchUtil, 'fetchResource');
        sinon.stub(FetchUtil, 'fetchDocumentation', () => Promise.resolve(Documentations.classWithOperation));
    });

    describe('loadResource', () => {

        it('should return object with matching @id', done => {
            FetchUtil.fetchResource.withArgs('http://example.com/resource')
                .returns(mockedResponse(Bodies.someJsonLd));

            Hydra.loadResource('http://example.com/resource')
                .then(res => {
                    expect(res['@id']).toBe('http://example.com/resource');
                    expect(res instanceof Resource).toBe(true);
                    done();
                })
                .catch(done.fail);
        });

        it('should return object with matching @id, trailing slash ignored', done => {
            FetchUtil.fetchResource.withArgs('http://example.com/resource/')
                .returns(mockedResponse(Bodies.someJsonLd));

            Hydra.loadResource('http://example.com/resource/')
                .then(res => {
                    expect(res['@id']).toBe('http://example.com/resource');
                    done();
                })
                .catch(done.fail);
        });

        it('should return object with matching @id, trailing slash ignored', done => {
            FetchUtil.fetchResource.withArgs('http://example.com/resource')
                .returns(mockedResponse(Bodies.idWithTrailingSlash));

            Hydra.loadResource('http://example.com/resource')
                .then(res => {
                    expect(res['@id']).toBe('http://example.com/resource/');
                    done();
                })
                .catch(done.fail);
        });

        it('should load documentation', done => {
            FetchUtil.fetchResource.withArgs('http://example.com/resource')
                .returns(mockedResponse(Bodies.someJsonLd));

            Hydra.loadResource('http://example.com/resource')
                .then(() => {
                    expect(FetchUtil.fetchDocumentation.calledWithMatch('http://api.example.com/doc/')).toBe(true);
                    done();
                })
                .catch(done.fail);
        });

        it('should turn JSON-LD into a graph of objects', done => {
            FetchUtil.fetchResource.withArgs('http://example.com/resource')
                .returns(mockedResponse(Bodies.someJsonLd));

            Hydra.loadResource('http://example.com/resource')
                .then(res => {
                    expect(Object.is(res['http://example.com/vocab#other'], res['http://example.com/vocab#other_yet'])).toBe(true);
                    expect(res['http://example.com/vocab#other']['@id']).toBe('http://example.com/linked');
                    done();
                })
                .catch(done.fail);
        });

        it('should turn object with arrays into matching object graph', done => {
            FetchUtil.fetchResource.withArgs('http://example.com/resource')
                .returns(mockedResponse(Bodies.hydraCollection));

            Hydra.loadResource('http://example.com/resource')
                .then(res => {
                    expect(res[Core.Vocab.member].length).toBe(4);
                    _.each(res[Core.Vocab.member], member => {
                        expect(member instanceof Resource).toBe(true);
                    });
                    done();
                })
                .catch(done.fail);
        });

        it('should make each nested object a Resource', done => {
            FetchUtil.fetchResource.withArgs('http://example.com/resource')
                .returns(mockedResponse(Bodies.hydraCollection));

            Hydra.loadResource('http://example.com/resource')
                .then(res => {
                    expect(res['http://example.vocab/managedBy'] instanceof Resource)
                        .toBe(true, 'was ' + JSON.stringify(res['http://example.vocab/managedBy']));
                    done();
                })
                .catch(done.fail);
        });

        it('should load parent of collection view as Resource', done => {
            FetchUtil.fetchResource.withArgs('http://example.com/resource?page=3')
                .returns(mockedResponse(Bodies.hydraCollectionWithView));

            Hydra.loadResource('http://example.com/resource?page=3')
                .then(res => {
                    expect(res.collection).toBeDefined();
                    expect(res.collection instanceof Resource).toBe(true, 'Actual type is: ' + res.collection.constructor.name);
                    done();
                })
                .catch(done.fail);
        });

        it('should fail when resource with given @id doesn\'t exist in the representation', done => {
            FetchUtil.fetchResource.withArgs('http://example.com/not/there')
                .returns(mockedResponse(Bodies.someJsonLd));

            Hydra.loadResource('http://example.com/not/there')
                .then(done.fail, err => {
                    expect(err.message).toBe('Resource http://example.com/not/there was not found in the response');
                    done();
                })
                .catch(done.fail);
        });

        it('should discover incoming links for resources', done => {
            FetchUtil.fetchResource.withArgs('http://example.com/resource')
                .returns(mockedResponse(Bodies.someJsonLd));

            Hydra.loadResource('http://example.com/resource')
                .then(res => {
                    expect(res['http://example.com/vocab#other'].incomingLinks.length).toBe(2);

                    var incomingLinks = res['http://example.com/vocab#other'].incomingLinks;
                    expect(_.some(incomingLinks, { subjectId: 'http://example.com/resource', predicate: 'http://example.com/vocab#other' })).toBe(true);
                    expect(_.some(incomingLinks, { subjectId: 'http://example.com/resource', predicate: 'http://example.com/vocab#other_yet' })).toBe(true);
                    done();
                })
                .catch(done.fail);
        });

        it('should pass each object through ResourceFactory', done => {
            FetchUtil.fetchResource.withArgs('http://example.com/resource')
                .returns(mockedResponse(Bodies.hydraCollection));

            Hydra.loadResource('http://example.com/resource')
                .then(() => {
                    var ids = _.map(Hydra.resourceFactory.createResource.getCalls(), call => {
                        return call.args[0]['@id'];
                    });
                    expect(Hydra.resourceFactory.createResource.callCount)
                        .toBe(6, 'Actual calls for: ' + ids);
                    done();
                })
                .catch(done.fail);
        });

        it('should load resource with deep blank node structure', done => {
            FetchUtil.fetchResource.withArgs('http://example.com/root')
                .returns(mockedResponse(Bodies.deepBlankNodes));

            Hydra.loadResource('http://example.com/root')
                .then(res => {
                    var p = 'http://example.com/prop';
                    var t = 'http://example.com/text';

                    expect(res[p][p][p][p][t]).toBe('I\'m nested way deep');
                    done();
                })
                .catch(done.fail);
        })
    });

    afterEach(() => FetchUtil.fetchResource.restore());
    afterEach(() => FetchUtil.fetchDocumentation.restore());
    afterEach(() => Hydra.resourceFactory.createResource.restore());

});

function mockedResponse(resource) {
    return jsonld.flatten(resource, {})
        .then(expanded => ({
            resources: expanded[JsonLd.Graph],
            apiDocumentationLink: 'http://api.example.com/doc/'
        }));
}