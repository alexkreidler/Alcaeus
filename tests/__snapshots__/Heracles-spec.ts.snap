// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Hydra loadDocumentation should replace its representation in the dataset when loading twice 1`] = `
"<http://api.example.com/doc/2> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://www.w3.org/ns/hydra/core#ApiDocumentation> <http://api.example.com/doc/> .
"
`;

exports[`Hydra loadDocumentation should store its representation in the dataset 1`] = `
"<http://api.example.com/doc/> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://www.w3.org/ns/hydra/core#ApiDocumentation> <http://api.example.com/doc/> .
"
`;
